### PR TITLE
Move in extension types

### PR DESCRIPTION
**Pull request by karelhala**
JIRA: [HAC-424](https://issues.redhat.com/browse/HAC-424)

* [x] core.action/provider
* [x] core.action/resource-provider
* [x] core.action/group
* [x] core.action/filter
* [x] core.catalog/item-type
* [x] core.catalog/item-provider
* [x] core.catalog/item-filter
* [x] core.context-provider
* [x] core.flag
* [x] core.flag/model
* [x] core.navigation/href
* [x] core.navigation/resource-ns
* [x] core.navigation/separator
* [x] core.page/resource/list
* [x] core.page/resource/details
* [x] core.redux-reducer
* [x] core.resource/create
* [x] core.resource-metadata
 * this type doesn't exists in console, there's just `console.model-metadata` I've exported this extension type and also `core.resource-metadata`, they are both identical. We should consider deprecating one or the other.
* [x] core.telemetry/listener
* [x] core.yaml-template

Additional extensions:
* [x] core.page/route

### CODE REVIEW

**Comment by karelhala** [comment]
ping @vojtechszocs

File: packages/lib-core/src/api/common-types.ts
Comment (andrewballantyne) [inline packages/lib-core/src/api/common-types.ts:9]: `ReturnType<ExtensionHook<...>>` will return this value. I think we can avoid exporting this.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Partial review... ran out of time. Will continue tomorrow.

File: packages/lib-core/src/api/common-types.ts
Comment (andrewballantyne) [reply packages/lib-core/src/api/common-types.ts]: It should see the [TypeScript Playground example](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggdiA8gIwFYQMbCgXigbwF8BuKAejKgDMB7AJymAgGdgBLOAcwBpGALNsygYAhnCjJoYkMAFcAsACglFKABVw0WgwgAPJnGZsa4vjRoBrJXrD1soSFACi+iIeNwAEuYsAeNRqQvGqIYOwmQnoGACZC8EhomNh48SjoWAB8uFAAFEpQUDRhHswAXOqh4YbcSgCUuFkuBkYm3pYASiwArgA2wP6BEBnESiqUg9T0UFFuLaY+UHTdffwi2EvAXXSGUADa0WsivEvMND0AbhDRvBB0dPQAuta6tnT2ms6u7q0+ncy9-QCmiyeF2QKCEnMPQgYl4XTgFjgNAA7nAHiNlJiHNAAMo0AC2EAAwl1WASJnh8FA4CJCeVWHQOJwoCRRopVABZGik6A0S4MAAGahYwAFjBoUGYEGgshlHzY2D4ImYSmx6hF2U6m22g18TVmHjafjxhJJZPxgwyGSAA) I coded up to test it.

File: packages/lib-core/src/api/common-types.ts
Comment (vojtechszocs) [reply packages/lib-core/src/api/common-types.ts:38]: @andrewballantyne AFAIK this was discussed with @spadgett and the consensus was to have a single way to identify k8s resources - check out `K8sResourceIdentifier` type in #30 - group is optional but version and kind are mandatory:

```ts
type K8sResourceIdentifier = {
  apiGroup?: string;
  apiVersion: string;
  kind: string;
};
```

Notice that k8s APIs use `apiGroup` instead of `group` and `apiVersion` instead of `version` - I wouldn't change that in `K8sResourceIdentifier` type, as they reflect the k8s API response data.

TL;DR - I'd suggest introducing a single type via e.g. `packages/lib-core/src/types/extension-k8s.ts` which is conceptually equivalent to above.

```ts
type ExtensionK8sResourceIdentifier = {
  group?: string;
  version: string;
  kind: string;
};
```

Comment (vojtechszocs) [reply packages/lib-core/src/api/common-types.ts:38]: It's technically possible to reuse `K8sResourceIdentifier` from `lib-utils` and rename the group/version property names (like [this](https://stackoverflow.com/a/52702528)) but that would introduce `lib-core` => `lib-utils` (or the opposite) dependency, which we generally want to avoid.

Core k8s types should be kept in `lib-utils` since this package deals primarily with k8s API calls, among other things.

File: packages/lib-core/src/api/common-types.ts
Comment (vojtechszocs) [reply packages/lib-core/src/api/common-types.ts]: The `React.ReactNode` part should definitely fail without importing React as `React`, this is weird.

File: packages/lib-core/src/api/common-types.ts
Comment (vojtechszocs) [reply packages/lib-core/src/api/common-types.ts]: We already have `UseResolvedExtensionsResult<TExtension extends Extension>` type declared in `packages/lib-core/src/store/useResolvedExtensions.ts` so we should reuse that one.

The copied-over `ExtensionHook` type actually refers to the result of `useResolvedExtensions` hook.

Comment (andrewballantyne) [reply packages/lib-core/src/api/common-types.ts]: I'm fine with it either way.

My thoughts are - I think we could just use the `ReturnType` for types that have no name-benefit. It's a 3-param array and is just a collection of responses. A name will not help us use it anymore than if it was a `ReturnType<...>` result.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/actions.ts:7]: FYI @andrewballantyne our `Extension` type has two generic type parameters, as opposed to `Extension` type in Console (can't change the latter easily since it's already used by a lot of code).

This eliminates the need for `ExtensionDeclaration` type in this SDK.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/actions.ts:8]: @andrewballantyne We should consider writing a custom ESLint rule that checks all specific extension types and ensures the first generic type parameter is a string literal that starts with `core.` prefix.

https://typescript-eslint.io/docs/development/custom-rules/

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/actions.ts:11]: TS coerces this to just `string` so I'm not sure why we need the `| 'resource'` part.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/actions.ts:14]: Nit: please sanitize the doc comments, ideally it should be either
```ts
/** one-line comment */
```
or
```ts
/**
 * multi-line comment
 * multi-line comment
 */
```

Note that JSDoc comments should always start with `/**` and not with `/*`.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/actions.ts:15]: I'll look into all `ExtensionHook<T>` type references as part of a second round of code review.

I'm not sure why we need to reference the signature of the extension consuming hook itself, I'd assume this is primarily due to type/signature commonality in which case we should introduce a type with a more suitable name.

@andrewballantyne FYI this needs looking into.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/actions.ts:43]: I guess `scope` was originally typed as `any` and replaced with `unknown` type.

Ideally we would provide a better type here, it should be more apparent from how the extension is used.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/actions.ts:49]: These type guards are technically boilerplate code, using the following template:

```
export const is${ExtensionTypeName} = (e: Extension): e is ${ExtensionTypeName} => {
  return e.type === ${ExtensionTypeLiteralValue};
};
```

I'd be happier if we could generate them to avoid having to maintain them manually.

@andrewballantyne We should explore using TypeScript custom transformers to generate those type guards during TS compilation. The Rollup plugin we're using `@rollup/plugin-typescript` already supports this TS compiler feature.

https://github.com/madou/typescript-transformer-handbook

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/actions.ts:67]: By convention, these are helper types which are relevant for / used in this file only.

File: packages/lib-core/src/extensions/core-types.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/core-types.ts:1]: Similar to my above comment, let's have a consistent way to represent all types instead of spreading them over the source tree.

@andrewballantyne If we identify an overlap / commonality between k8s `lib-core` types vs. k8s `lib-utils` types, we can move them into the `common` package, whose code & types is bundled into all libraries that use it.

File: packages/lib-core/src/extensions/pages.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/pages.ts:13]: @andrewballantyne FYI the `& AnyObject` was a workaround related to JSON schema generation via [ts-json-schema-generator](https://github.com/vega/ts-json-schema-generator) tool.

I think we should remove this workaround here and on other places.

**Review by vojtechszocs — CHANGES_REQUESTED** [review]

File: packages/lib-core/package.json
Comment (karelhala) [reply packages/lib-core/package.json:46]: Good point! I'll move the redux to devDeps and peerDeps as for the yup and lodash-es I'd rather not make this PR bigger than it needs to be.

File: packages/lib-core/src/api/common-types.ts
Comment (karelhala) [reply packages/lib-core/src/api/common-types.ts:15]: Usually the `title` is optional. I was considering making it optional in here as well and for those usecases where the title was required to extend it. But that was less often, than keeping the `title` required (only one other place had it optional).

This type was introduced solely in this PR, as I said it felt a bit better to have it as common type than to write 5 times the same type signature.

File: packages/lib-core/src/api/common-types.ts
Comment (karelhala) [reply packages/lib-core/src/api/common-types.ts:34]: I'll mark all as optional and where some will be required I'll mark them as optional. What about backwards compatibility? Should we export the type under multiple values?

Comment (karelhala) [reply packages/lib-core/src/api/common-types.ts:34]: A bit better approach, I've added `ExtensionK8sResourceIdentifier` which has all values optional and kept the previous types just extending this one.

File: packages/lib-core/src/extensions/actions.ts
Comment (karelhala) [reply packages/lib-core/src/extensions/actions.ts:1]: Mhh, that won't really work for types. With the above import `import * as React from 'react'` it is throwing an error of `All imports in the declaration are only used as types. Use 'import type'.` I can turn that rule off, but it feels a bit clunky to import every React type just to use one of them.

File: packages/lib-core/src/extensions/actions.ts
Comment (karelhala) [reply packages/lib-core/src/extensions/actions.ts:11]: Yeah, this felt a bit funky when I moved it over. But I thought it had some purpose when type suggestion. Removing the `| 'resource'` part!

File: packages/lib-core/src/extensions/actions.ts
Comment (karelhala) [reply packages/lib-core/src/extensions/actions.ts:14]: Yeah, good point!

File: packages/lib-core/src/extensions/actions.ts
Comment (karelhala) [reply packages/lib-core/src/extensions/actions.ts:43]: I'm not sure I can provide much information what the `scope` can be here :disappointed:

File: packages/lib-core/src/extensions/pages.ts
Comment (karelhala) [reply packages/lib-core/src/extensions/pages.ts:13]: We're using the `AnyObject` in frontend-components repository to consume additional props. Could be usefull in these extension definitionas (if someone implements them they can add custom props as well). But if we want to prevent this let's nuke the usage of `AnyObject` here.

File: packages/lib-core/src/extensions/actions.ts
Comment (christianvogt) [reply packages/lib-core/src/extensions/actions.ts:48]: The `scope` is different for every context. For example in topology the `scope` would be a scene graph element, while for a resource page, the scope would be the resource object itself.
The intent was for reusing the same extension for supplying actions in different contexts. But maybe it makes more sense to use completely different extension types that inherit the same structure ?

File: packages/lib-core/src/types/core.ts
Comment (andrewballantyne) [inline packages/lib-core/src/types/core.ts:18]: I feel this is an "OpenShift Console"-esk of a type and not an accurate type. We had a serious problem (less so now) about types in OpenShift Console being fully optional... making our code a mess of if-statements.

May need a bit of investigation 🤔

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: packages/lib-core/src/extensions/actions.ts
Comment (christianvogt) [reply packages/lib-core/src/extensions/actions.ts:48]: `scope` should follow the same typing as `ExtensionHook` which accepts a typed parameter in the function call.
Unsure if `unknown` or `any` is more appropriate here but the type can be cast to a specific type based on the `contextId`.

Comment (christianvogt) [reply packages/lib-core/src/extensions/actions.ts:48]: resource actions will be important to bring over for supporting details and list pages which present resource actions to the user.

Comment (karelhala) [reply packages/lib-core/src/extensions/actions.ts:48]: So the function type can be written as:
```TS
type FilterType<TType> = (scope: TType, action: Action) => boolean
```

File: packages/lib-core/src/types/extension.ts
Comment (karelhala) [reply packages/lib-core/src/types/extension.ts:4]: Ah, yeah. This was my blind attempt to change the type for `Extension` as suggested in here https://github.com/openshift/dynamic-plugin-sdk/pull/29#discussion_r804916683 will rol back to the previous one.

File: packages/lib-core/src/types/common.ts
Comment (karelhala) [reply packages/lib-core/src/types/common.ts:22]: Since this was copied over from console I thought it would be nice to keep them so we don't break the apps when migrating to the new SDK.

File: packages/lib-core/src/types/common.ts
Comment (karelhala) [reply packages/lib-core/src/types/common.ts:15]: There were 3 places where the extension had `type` and `title` I didn't wanted to blindly copy them over just out of spite. We could in theory keep them in `catalog` extension (and not export it) as that's the place where this type is used.

File: packages/lib-core/src/extensions/actions.ts
Comment (andrewballantyne) [reply packages/lib-core/src/extensions/actions.ts:48]: Not sure the type necessarily helps in this case. The user is a plugin. They define their functionality -- the caller is the App (in this case) and they will provide what they provide based on where they call it.

The Generic type won't translate well between the two I don't think. It's a weird concept. Not sure why we have such a loose extension. I think we can leave it as `unknown` for now imo. The plugin can type it when they know the use-case they built it for for their own internal function typing.

These action types are weird 😕

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [reply packages/lib-core/src/extensions/actions.ts:1]: Oh right, @karelhala we should use `import type` whenever all imported variables refer to types.

When we need to import both some types and values from some module, there would be two import statements, one `import` and one `import type`.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [reply packages/lib-core/src/extensions/actions.ts:7]: Right, we don't need two types because we have the ability to change the base `Extension` one :smiley: in other words, `Extension` with `type: string` in Console is a legacy bit that we can mitigate here.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [reply packages/lib-core/src/extensions/actions.ts:16]: > as part of a second round of code review

Can be done after merging this PR, I don't want to hold it on this.

@andrewballantyne My concern was why do we even need to know the signature of `useResolvedExtension` hook (function) when dealing with extension-related types :smiley: is it because of structural similarity, or other reason? As I wrote, we can improve this later on.

Comment (vojtechszocs) [reply packages/lib-core/src/extensions/actions.ts:16]: To elaborate, extension-related types (i.e. ones referenced by specific `Extension` declarations) are used to _describe_ specific extension points in the application; basically using data and/or callback functions.

I'm missing the logical link to `useResolvedExtension` which is a function used to consume/interpret extensions.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [reply packages/lib-core/src/extensions/actions.ts:48]: > The Generic type won't translate well between the two I don't think.

Right, if it's generic, both app and plugin need to make an assumption about it.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [reply packages/lib-core/src/extensions/actions.ts:54]: This can also be handled as part of [HAC-376](https://issues.redhat.com/browse/HAC-376).

File: packages/lib-core/src/extensions/pages.ts
Comment (vojtechszocs) [reply packages/lib-core/src/extensions/pages.ts:13]: I think the default should be to explicitly list all possible properties (mandatory and/or optional) and disallow unknown/custom ones.

There may be use cases where we can allow that, but I think it shouldn't be the default pattern.

File: packages/lib-core/src/types/extension.ts
Comment (vojtechszocs) [reply packages/lib-core/src/types/extension.ts]: Yup, extension's `type` property should always be `string` at runtime, but here we allow it to be narrowed-down to a specific literal type. (In TypeScript, `'foo'` type is assignable to `string` type, but not the other way around.)

File: packages/lib-core/src/types/common.ts
Comment (vojtechszocs) [reply packages/lib-core/src/types/common.ts:22]: Right, let's not bring in unnecessary complexity if possible.

File: .gitignore
Comment (vojtechszocs) [inline .gitignore:16]: I think this is out of scope for this PR :smiley:

File: packages/lib-core/package.json
Comment (vojtechszocs) [inline packages/lib-core/package.json:36]: Any reason not to use a semver pattern, but a specific version? (Just asking, I'm OK with this change.)

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [reply packages/lib-core/src/extensions/actions.ts:16]: OK just realized it's because we want to pass React hook via `provider` property.

File: packages/lib-core/src/extensions/actions.ts
Comment (vojtechszocs) [reply packages/lib-core/src/extensions/actions.ts:48]: I feel it's still better typed with a generic type param (falling back to `unknown`) but we can add it later.

File: packages/lib-core/src/extensions/pages.ts
Comment (vojtechszocs) [reply packages/lib-core/src/extensions/pages.ts:13]: Let's nuke `& AnyObject` part for now, its original purpose was to make https://github.com/vega/ts-json-schema-generator happy in context of Console plugin SDK.

File: packages/lib-core/src/extensions/pages.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/pages.ts:19]: Let's nuke `& AnyObject` part for now, its original purpose was to make https://github.com/vega/ts-json-schema-generator happy in context of Console plugin SDK.

File: packages/lib-core/src/types/common.ts
Comment (vojtechszocs) [inline packages/lib-core/src/types/common.ts:7]: TypeScript 4+ allows tuple elements to be named, we already do that in tuple type `UseResolvedExtensionsResult`. We should improve this in a follow-up change.

File: packages/lib-core/src/types/common.ts
Comment (vojtechszocs) [inline packages/lib-core/src/types/common.ts:24]: We can create mapped type helpers, for example:

```ts
type Bar = RequiredProperties<Foo, 'a' | 'b'>; // remove optional modifier for given properties of Foo
type Bar = OptionalProperties<Foo, 'a' | 'b'>; // add optional modifier for given properties of Foo
```

Can be done later as follow-up improvement.

File: packages/lib-core/src/types/core.ts
Comment (vojtechszocs) [reply packages/lib-core/src/types/core.ts:18]: Ping, if we don't need `AccessReviewResourceAttributes` let's remove it for now.

**Review by vojtechszocs — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (karelhala) [reply .gitignore:16]: yeah :smile:

File: packages/lib-core/package.json
Comment (karelhala) [reply packages/lib-core/package.json:36]: The router v6 is not backwards compatible. But good point about adding carret in this one.

File: packages/lib-core/src/extensions/pages.ts
Comment (karelhala) [reply packages/lib-core/src/extensions/pages.ts:13]: Sure thing!

File: packages/lib-core/src/extensions/pages.ts
Comment (karelhala) [reply packages/lib-core/src/extensions/pages.ts:19]: Removing it.

File: packages/lib-core/src/types/common.ts
Comment (karelhala) [reply packages/lib-core/src/types/common.ts:7]: Yeah, that makes sense. I'll keep as is, and we can iterate on it.

File: packages/lib-core/src/types/core.ts
Comment (karelhala) [reply packages/lib-core/src/types/core.ts:18]: It's used in `actions.ts` I'll remove the optional prop that is typed by this.

**Review by andrewballantyne — APPROVED** [review]
Looks good to me. Waiting on @vojtechszocs final review & the test passes 🙂

File: packages/lib-core/src/extensions/catalog.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/catalog.ts:59]:
```suggestion
export type CatalogItem<TData extends AnyObject = AnyObject> = {
```

File: packages/lib-core/src/extensions/context-providers.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/context-providers.ts:5]:
```suggestion
export type ContextProvider<TValue = unknown> = Extension<
```

File: packages/lib-core/src/types/common.ts
Comment (vojtechszocs) [inline packages/lib-core/src/types/common.ts:4]:
```suggestion
export type ExtensionHook<TResult, TOptions extends AnyObject = AnyObject> = (
```

File: packages/lib-core/src/types/common.ts
Comment (vojtechszocs) [inline packages/lib-core/src/types/common.ts:8]: Please add the following comment above:
```
// TODO(vojtech): use types like RequiredProperties<T, K> and OptionalProperties<T, K>
// to change optionality of specific properties, instead of redefining them via "&" intersection
```

File: packages/lib-core/src/extensions/model-metadata.ts
Comment (vojtechszocs) [inline packages/lib-core/src/extensions/model-metadata.ts:12]:
```suggestion
// Arbitrary types
```

**Review by vojtechszocs — CHANGES_REQUESTED** [review]

File: packages/lib-core/src/store/useResolvedExtensions.ts
Comment (karelhala) [reply packages/lib-core/src/store/useResolvedExtensions.ts:15]: Yup, good point!

**Pull request merged**
